docs/design_notes/module_lifecycle.md

```diff
@@ -0,0 +1,289 @@
+
+Another benefit of defining submodules and/or variables inline is that you can use arguments to your method when constructing variables. The most common example of this is using shape information to determine the shape of a parameter like this:
+
+
+```python
+class CompactScaledMLP(nn.Module):
+  hidden_size: int
+  out_size: int
+
+  @nn.compact
+  def __call__(self, x):
+    scale = self.param("scale", x.shape[-1:], nn.initializers.ones)
+    x *= scale
+    a = nn.Dense(self.hidden_size)(x)
+    h = nn.relu(a)
+    return nn.Dense(self.out_size)(h)
+```
+
+
+Many of the standard linen layers like `nn.Dense` use this type of shape inference already to avoid the need to specify input shapes (like the number of input features to a Dense layer).
+
+### Compact control flow
+
+The order in which you define submodules determines the name of a submodule if none is provided explicitly (using the name= construction argument). Because the name determines how parameters are mapped to submodules, you must be careful about mixing control flow with auto-generated names. Using control flow can change the order or remove certain submodules altogether. This is useful in the case where certain submodules should only exist depending on some construction argument. However, when control flow depends on the input arguments to the Module you should be careful. For example the following Module will break:
+
+
+```python
+class WrongModule(nn.Module):
+  @nn.compact
+  def __call__(self, x, mode):
+    if mode == "encode"
+      return nn.Dense(8)(x)
+    elif mode == "decode":
+      return nn.Dense(4)(x)
+```
+
+
+This problem can be solved in various ways: Provide explicit names, create the modules in setup, or move the constructor out of the control flow. The latter is done as follows:
+
+
+```python
+class CorrectModule(nn.Module):
+  @nn.compact
+  def __call__(self, x, mode):
+    encoder = nn.Dense(8)
+    decoder = nn.Dense(4)
+    if mode == "encode"
+      return encoder(x)
+    elif mode == "decode":
+      return decoder(x)
+```
+
+
+In the above example the construction order is fixed. After construction the submodules can be used in an arbitrary order.
+
+> Note \
+compact modules show a strong resemblance to [React hooks](https://reactjs.org/docs/hooks-custom.html)
+
+
+## Top-level Modules
+
+When a Module instance is created at the top-level it will be in an "unbound" state, meaning it has no variables attached. By top-level we mean it is not constructed as a sub-Module inside another Module class. Apart from calling `init` and `apply` there is not much you can do with an unbound Module. Note also that `setup` is not called on unbound Modules so you can really only access the construction arguments. See the Future work section for how this might change in the future.
```

Comment:
```suggestion
When a Module instance is created at the top level it will be in an "unbound" state, meaning it has no variables attached. By top-level we mean it is not constructed as a sub-Module inside another Module class. Apart from calling `init` and `apply` there is not much you can do with an unbound Module. Note also that `setup` is not called on unbound Modules so you can really only access the construction arguments. See the Future work section for how this might change in the future.
```

Reply:
A modification of @marcvanzee 's suggestion with some formatting:

```suggestion
When a Module instance is created at the "top-level", it will be in an "unbound" state—that is, it has no variables attached. "Top-level" means it is not constructed as a sub-Module inside another Module class. Apart from calling `init` and `apply`, there is not much you can do with an unbound Module. Note also that `setup` is not called on unbound Modules, so you can really only access the construction arguments. Refer to the **Future work** section to learn how this might change in the future.
```

NOTE: "sub-Module" or "submodule" - should pick one for consistency across all docs IMO @jheek @marcvanzee @avital